### ISSUE TITLE
Characters from Korean, Chinese, etc don’t show in txt/md file previews

### ISSUE THREAD

**Issue by HuidaeCho** [nextcloud/server#4198]
### Steps to reproduce
1. Create a text file in Korean
2. Preview image has broken fonts because the default OpenSans-Regular.ttf has no Korean fonts

### Expected behaviour
There should be a way to specify a user font file for TXT preview in config.php. Something like $CONFIG['fontFile']

### Actual behaviour
Broken text in preview or I have to modify lib/private/Preview/TXT.php to point $fontFile to a user font file.

**Comment by MorrisJobke** [comment]
I would vote for no config option, but a font that actually support Korean and Latin characters.

cc @georgehrke @jancborchardt

**Comment by jancborchardt** [comment]
Can we add a fallback font to the preview generator? In case characters don't exist, it should simply use Liberation Sans for example.

cc @georgehrke @rullzer @icewind1991

**Comment by skjnldsv** [comment]
Similar of #13377

**Comment by jancborchardt** [comment]
It’s not similar, as #13377 is about the interface fonts, and there it uses fallback fonts which are perfectly readable – just not as nice.

But for the previews, it shows _no_ fallback font, but rectangle boxes. (Just like for emojis etc.)

Here is the file in the text editor, perfectly fine:

![special characters file](https://user-images.githubusercontent.com/925062/59367814-2b456380-8d3d-11e9-8110-40f39625ee7d.png)

And here its preview, broken:
![special characters file preview](https://user-images.githubusercontent.com/925062/59367813-2aaccd00-8d3d-11e9-9c36-df5203d9ff62.png)

**Comment by skjnldsv** [comment]
Well, adding more compatible fonts to the system like #13377 suggest and to the preview generator will fix it, so yes, this is similar.

**Comment by skjnldsv** [comment]
@jancborchardt Shall we add this into the todo of the other issue?

**Comment by jancborchardt** [comment]
@skjnldsv ok, if you say it’s heavily related we can add it as todo of the other issue.

**Comment by skjnldsv** [comment]
@jancborchardt Well, it make sense to me, but I might be missing some keys here :wink: :see_no_evil: 
Let's leave this one opened as even if one is also the solution to the other, we'll not forget there are two parts to the problem! :smile_cat:

**Comment by jancborchardt** [comment]
Just as a note, this is _not_ fixed by using a native font stack as per https://github.com/nextcloud/server/pull/16055 as this is only for the interface fonts.

For the preview generator we would need TTF files for all the different fonts (as I guess WOFF2 is not supported by the preview generator?) And that means loading _a lot_ of files for Korean, Chinese and Japanese:

These would be the woff2 files, 184 for both regular and bold: https://fonts.googleapis.com/css?family=Noto+Sans+KR:400,700&display=swap

@rullzer is this doable somehow?

**Comment by Kellermaan** [comment]
Hi, I'm not sure this issue have been fixed.
I'm using the latest container version of nextcloud（Nextcloud 20.0.1）and I find out the preview of markdown file got not Chinese character. really strange.
The picture 1 is the real content and the picture 2 is the preview.
![微信截图_20201210001507](https://user-images.githubusercontent.com/11674190/101656036-36399b00-3a7d-11eb-90cb-7a4127d5224b.png)
![微信截图_20201210001500](https://user-images.githubusercontent.com/11674190/101655984-2d48c980-3a7d-11eb-81fb-064ed167fb2f.png)

Hope someone can tell me how to fix it, thank very much.
@jancborchardt

**Comment by Kellermaan** [comment]
can anyone help me on this? @rullzer @MorrisJobke @jancborchardt @HuidaeCho

**Comment by HuidaeCho** [comment]
> can anyone help me on this? @rullzer @MorrisJobke @jancborchardt @HuidaeCho

It may have changed now, but for version 13.0.12, I added one line after line 72:
```php
                $fontFile  = __DIR__;
                $fontFile .= '/../../../core';
                $fontFile .= '/fonts/OpenSans-Regular.ttf'; # line 72
                # by Huidae Cho
                $fontFile = '/usr/local/share/fonts/TTF/UnDotum.ttf'; # here, specify your own font path
```

**Comment by huww98** [comment]
Fonts that support CJK characters can be 20MB each. Is it acceptable to add these fonts to `/core/fonts` in this repo? If so, I can try to fix this, (also #13376). If not, we may need to add a config that allows us to choose our own font file.

**Comment by ludishur** [comment]
Thank you @HuidaeCho.
In my case version 20.0.2, I changed more.

```
lib/private/Avatar/Avatar.php:152:			$font = __DIR__ . '/../../core/fonts/NotoSans-Regular.ttf';
lib/private/Avatar/Avatar.php:187:		$font = __DIR__ . '/../../../core/fonts/NotoSans-Regular.ttf';
```

**Comment by Kellermaan** [comment]
@HuidaeCho @ludishur   Hi, I'm not so familiar with php program. when you said change the code. and then what? just restart the nextcloud container？

**Comment by Kellermaan** [comment]
@HuidaeCho @ludishur   I changed the code and restart the container, It seems didn't work. I also copied the ttf file into the dir the code mention。

**Comment by ludishur** [comment]
@Kellermaan 
There is no need to restart container after modifying the code.
You need to change username to see the code in action. 
User can change user name in settings menu.
Admin can be renamed in the user/group menu.
The avatar png file seems to be recreated when the username is changed.
The browser cache may remain, so it is recommended to open the secret mode browser and check it.
I hope it goes well.

**Comment by Kellermaan** [comment]
@ludishur  For me, the problem is the preview of md file can now show the chinese character. And I tried your method, it didn't work, am I missing some configuration? 
Thanks for your reply!!

**Comment by Kellermaan** [comment]
@ludishur  I fixed the issue, it turnes out that the default noto font doesn't include Chinese character. So I downloaded one include Chinese Japaness. Now it worked, thanks anyway.

**Comment by szaimen** [comment]
Hi, please update to at least 23.0.12 and report back if it fixes the issue. Thank you!

**Comment by joshtrichards** [comment]
Re-opened because as far as I can tell CJK support was never added for MD/TXT previews/thumbnails. Triggered by a post on the forum and some code review on my part.

There was discussion about it in #16040 and elsewhere, but the bundled fonts in `core/fonts` don't include the CJK (and other) variants of Noto as far as I can tell.

**Comment by jancborchardt** [comment]
@joshtrichards good catch. Do you know what needs to be done and/or could submit a pull request?

**Comment by jancborchardt** [comment]
(Also cc @juliushaertl regarding the previews)

**Comment by jancborchardt** [comment]
It’s probably best if we switch to Noto for this as it has the broadest support for various alphabets: https://fonts.google.com/noto/fonts?query=noto

**Comment by jancborchardt** [comment]
FYI @sorbaugh for planning

**Comment by susnux** [comment]
> It’s probably best if we switch to Noto

@jancborchardt we already did (or more specifically you did in 2019). The problem is that CJK fonts are too big and thus split for different languages and gd only supports rendering one line of text with one font.
